docs/env_variables.md
# Environmental variables

  This document discusses the environment variables used by American Fuzzy Lop++
  to expose various exotic functions that may be (rarely) useful for power
  users or for some types of custom fuzzing setups. See [README.md](README.md) for the general
  instruction manual.

## 1) Settings for all compilers

Starting with afl++ 3.0 there is only one compiler: afl-cc
To select the different instrumentation modes this can be done by
  1. passing the --afl-MODE command line option to the compiler
  2. or using a symlink to afl-cc: afl-gcc, afl-g++, afl-clang, afl-clang++,
     afl-clang-fast, afl-clang-fast++, afl-clang-lto, afl-clang-lto++,
     afl-gcc-fast, afl-g++-fast
  3. or using the environment variable `AFL_CC_COMPILER` with `MODE`

`MODE` can be one of `LTO` (afl-clang-lto*), `LLVM` (afl-clang-fast*), `GCC_PLUGIN`
(afl-g*-fast) or `GCC` (afl-gcc/afl-g++).


Because (with the exception of the --afl-MODE command line option) the
compile-time tools do not accept afl specific command-line options, they
make fairly broad use of environmental variables instead:

  - Most afl tools do not print any output if stdout/stderr are redirected.
    If you want to get the output into a file then set the `AFL_DEBUG`
    environment variable.
    This is sadly necessary for various build processes which fail otherwise.

  - Setting `AFL_HARDEN` automatically adds code hardening options when invoking
    the downstream compiler. This currently includes `-D_FORTIFY_SOURCE=2` and
    `-fstack-protector-all`. The setting is useful for catching non-crashing
    memory bugs at the expense of a very slight (sub-5%) performance loss.

  - By default, the wrapper appends `-O3` to optimize builds. Very rarely, this
    will cause problems in programs built with -Werror, simply because `-O3`
    enables more thorough code analysis and can spew out additional warnings.
    To disable optimizations, set `AFL_DONT_OPTIMIZE`.
    However if `-O...` and/or `-fno-unroll-loops` are set, these are not
    overridden.

  - Setting `AFL_USE_ASAN` automatically enables ASAN, provided that your
    compiler supports that. Note that fuzzing with ASAN is mildly challenging
    - see [notes_for_asan.md](notes_for_asan.md).

    (You can also enable MSAN via `AFL_USE_MSAN`; ASAN and MSAN come with the
    same gotchas; the modes are mutually exclusive. UBSAN can be enabled
    similarly by setting the environment variable `AFL_USE_UBSAN=1`. Finally
    there is the Control Flow Integrity sanitizer that can be activated by
    `AFL_USE_CFISAN=1`)

  - Setting `AFL_CC`, `AFL_CXX`, and `AFL_AS` lets you use alternate downstream
    compilation tools, rather than the default 'clang', 'gcc', or 'as' binaries
    in your `$PATH`.

  - `AFL_PATH` can be used to point afl-gcc to an alternate location of afl-as.
    One possible use of this is examples/clang_asm_normalize/, which lets
    you instrument hand-written assembly when compiling clang code by plugging
    a normalizer into the chain. (There is no equivalent feature for GCC.)

  - Setting `AFL_INST_RATIO` to a percentage between 0 and 100 controls the
    probability of instrumenting every branch. This is (very rarely) useful
    when dealing with exceptionally complex programs that saturate the output
    bitmap. Examples include v8, ffmpeg, and perl.

    (If this ever happens, afl-fuzz will warn you ahead of the time by
    displaying the "bitmap density" field in fiery red.)

    Setting `AFL_INST_RATIO` to 0 is a valid choice. This will instrument only
    the transitions between function entry points, but not individual branches.

    Note that this is an outdated variable. A few instances (e.g. afl-gcc)
    still support these, but state-of-the-art (e.g. LLVM LTO and LLVM PCGUARD)
    do not need this.

  - `AFL_NO_BUILTIN` causes the compiler to generate code suitable for use with
    libtokencap.so (but perhaps running a bit slower than without the flag).

  - `TMPDIR` is used by afl-as for temporary files; if this variable is not set,
    the tool defaults to /tmp.

  - If you are a weird person that wants to compile and instrument asm
    text files then use the `AFL_AS_FORCE_INSTRUMENT` variable:
      `AFL_AS_FORCE_INSTRUMENT=1 afl-gcc foo.s -o foo`

  - Setting `AFL_QUIET` will prevent afl-cc and afl-as banners from being
    displayed during compilation, in case you find them distracting.

  - Setting `AFL_CAL_FAST` will speed up the initial calibration, if the
    application is very slow.

## 2) Settings for LLVM and LTO: afl-clang-fast / afl-clang-fast++ / afl-clang-lto / afl-clang-lto++

The native instrumentation helpers (instrumentation and gcc_plugin) accept a subset
of the settings discussed in section 1, with the exception of:

  - LLVM modes support `AFL_LLVM_DICT2FILE=/absolute/path/file.txt` which will
    write all constant string comparisons  to this file to be used later with
    afl-fuzz' `-x` option.

  - `AFL_AS`, since this toolchain does not directly invoke GNU as.

  - `TMPDIR` and `AFL_KEEP_ASSEMBLY`, since no temporary assembly files are
    created.

  - `AFL_INST_RATIO`, as we by default use collision free instrumentation.
    Not all passes support this option though as it is an outdated feature.

Then there are a few specific features that are only available in instrumentation mode:

### Select the instrumentation mode

    - `AFL_LLVM_INSTRUMENT` - this configures the instrumentation mode. 
      Available options:
        CLASSIC - classic AFL (map[cur_loc ^ prev_loc >> 1]++) (default)
        CFG - InsTrim instrumentation (see below)
        LTO - LTO instrumentation (see below)
        CTX - context sensitive instrumentation (see below)
        NGRAM-x - deeper previous location coverage (from NGRAM-2 up to NGRAM-16)
      In CLASSIC (default) and CFG/INSTRIM you can also specify CTX and/or
      NGRAM, seperate the options with a comma "," then, e.g.:
        `AFL_LLVM_INSTRUMENT=CFG,CTX,NGRAM-4`
      Not that this is a good idea to use both CTX and NGRAM :)

### LTO

  This is a different kind way of instrumentation: first it compiles all
    code in LTO (link time optimization) and then performs an edge inserting
    instrumentation which is 100% collision free (collisions are a big issue
    in afl and afl-like instrumentations). This is performed by using
    afl-clang-lto/afl-clang-lto++ instead of afl-clang-fast, but is only
    built if LLVM 11 or newer is used.

   - `AFL_LLVM_INSTRUMENT=CFG` will use Control Flow Graph instrumentation.
     (not recommended for afl-clang-fast, default for afl-clang-lto as there
      it is a different and better kind of instrumentation.)

  None of the following options are necessary to be used and are rather for
    manual use (which only ever the author of this LTO implementation will use).
    These are used if several separated instrumentations are performed which
    are then later combined.

   - `AFL_LLVM_DOCUMENT_IDS=file` will document to a file which edge ID was given
     to which function. This helps to identify functions with variable bytes
     or which functions were touched by an input.
   - `AFL_LLVM_MAP_ADDR` sets the fixed map address to a different address than
     the default `0x10000`. A value of 0 or empty sets the map address to be
     dynamic (the original afl way, which is slower)
   - `AFL_LLVM_MAP_DYNAMIC` sets the shared memory address to be dynamic
   - `AFL_LLVM_LTO_STARTID` sets the starting location ID for the instrumentation.
     This defaults to 1
   - `AFL_LLVM_LTO_DONTWRITEID` prevents that the highest location ID written
     into the instrumentation is set in a global variable

  See [instrumentation/README.lto.md](../instrumentation/README.lto.md) for more information.

### INSTRIM

  This feature increases the speed by ~15% without any disadvantages to the
    classic instrumentation.

  Note that there is also an LTO version (if you have llvm 11 or higher) -
    that is the best instrumentation we have. Use `afl-clang-lto` to activate.
    The InsTrim LTO version additionally has all the options and features of
    LTO (see above).

   - Setting `AFL_LLVM_INSTRIM` or `AFL_LLVM_INSTRUMENT=CFG` activates this mode

   - Setting `AFL_LLVM_INSTRIM_LOOPHEAD=1` expands on INSTRIM to optimize loops.
      afl-fuzz will only be able to see the path the loop took, but not how
      many times it was called (unless it is a complex loop).

  See [instrumentation/README.instrim.md](../instrumentation/README.instrim.md)

### NGRAM

   - Setting `AFL_LLVM_NGRAM_SIZE` or `AFL_LLVM_INSTRUMENT=NGRAM-{value}`
      activates ngram prev_loc coverage, good values are 2, 4 or 8
      (any value between 2 and 16 is valid).
      It is highly recommended to increase the `MAP_SIZE_POW2` definition in
      config.h to at least 18 and maybe up to 20 for this as otherwise too
      many map collisions occur.

  See [instrumentation/README.ngram.md](../instrumentation/README.ngram.md)

### CTX

   - Setting `AFL_LLVM_CTX` or `AFL_LLVM_INSTRUMENT=CTX`
      activates context sensitive branch coverage - meaning that each edge
      is additionally combined with its caller.
      It is highly recommended to increase the `MAP_SIZE_POW2` definition in
      config.h to at least 18 and maybe up to 20 for this as otherwise too
      many map collisions occur.

  See [instrumentation/README.ctx.md](../instrumentation/README.ctx.md)

### LAF-INTEL

  This great feature will split compares into series of single byte comparisons
    to allow afl-fuzz to find otherwise rather impossible paths. It is not
    restricted to Intel CPUs ;-)

   - Setting `AFL_LLVM_LAF_TRANSFORM_COMPARES` will split string compare functions

   - Setting `AFL_LLVM_LAF_SPLIT_SWITCHES` will split all `switch` constructs

   - Setting `AFL_LLVM_LAF_SPLIT_COMPARES` will split all floating point and
      64, 32 and 16 bit integer CMP instructions

   - Setting `AFL_LLVM_LAF_SPLIT_FLOATS` will split floating points, needs
      AFL_LLVM_LAF_SPLIT_COMPARES to be set

   - Setting `AFL_LLVM_LAF_ALL` sets all of the above

  See [instrumentation/README.laf-intel.md](../instrumentation/README.laf-intel.md) for more information.

### INSTRUMENT LIST (selectively instrument files and functions)

  This feature allows selective instrumentation of the source

   - Setting `AFL_LLVM_ALLOWLIST` or `AFL_LLVM_DENYLIST` with a filenames and/or
      function will only instrument (or skip) those files that match the names
      listed in the specified file.

  See [instrumentation/README.instrument_list.md](../instrumentation/README.instrument_list.md) for more information.

### NOT_ZERO

   - Setting `AFL_LLVM_NOT_ZERO=1` during compilation will use counters
      that skip zero on overflow. This is the default for llvm >= 9,
      however for llvm versions below that this will increase an unnecessary
      slowdown due a performance issue that is only fixed in llvm 9+.
      This feature increases path discovery by a little bit.

   - Setting `AFL_LLVM_SKIP_NEVERZERO=1` will not implement the skip zero
      test. If the target performs only few loops then this will give a
      small performance boost.

  See [instrumentation/README.neverzero.md](../instrumentation/README.neverzero.md)

### CMPLOG

   - Setting `AFL_LLVM_CMPLOG=1` during compilation will tell afl-clang-fast to
      produce a CmpLog binary.

  See [instrumentation/README.cmplog.md](../instrumentation/README.cmplog.md)

## 3) Settings for GCC / GCC_PLUGIN modes

Then there are a few specific features that are only available in GCC and
GCC_PLUGIN mode.

  - Setting `AFL_KEEP_ASSEMBLY` prevents afl-as from deleting instrumented
    assembly files. Useful for troubleshooting problems or understanding how
    the tool works. (GCC mode only)
    To get them in a predictable place, try something like:
```
    mkdir assembly_here
    TMPDIR=$PWD/assembly_here AFL_KEEP_ASSEMBLY=1 make clean all
```
  - Setting `AFL_GCC_INSTRUMENT_FILE` with a filename will only instrument those
    files that match the names listed in this file (one filename per line).
    See [instrumentation/README.instrument_list.md](../instrumentation/README.instrument_list.md) for more information.
    (GCC_PLUGIN mode only)

## 4) Settings for afl-fuzz

The main fuzzer binary accepts several options that disable a couple of sanity
checks or alter some of the more exotic semantics of the tool:

  - Setting `AFL_SKIP_CPUFREQ` skips the check for CPU scaling policy. This is
    useful if you can't change the defaults (e.g., no root access to the
    system) and are OK with some performance loss.

  - `AFL_EXIT_WHEN_DONE` causes afl-fuzz to terminate when all existing paths
    have been fuzzed and there were no new finds for a while. This would be
    normally indicated by the cycle counter in the UI turning green. May be
    convenient for some types of automated jobs.

  - `AFL_MAP_SIZE` sets the size of the shared map that afl-fuzz, afl-showmap,
    afl-tmin and afl-analyze create to gather instrumentation data from
    the target. This must be equal or larger than the size the target was
    compiled with.

  - `AFL_TESTCACHE_SIZE` allows you to override the size of `#define TESTCASE_CACHE`
    in config.h. Recommended values are 50-250MB - or more if your fuzzing
    finds a huge amount of paths for large inputs.

  - Setting `AFL_DISABLE_TRIM` tells afl-fuzz not to trim test cases. This is
    usually a bad idea!

  - Setting `AFL_NO_AFFINITY` disables attempts to bind to a specific CPU core
    on Linux systems. This slows things down, but lets you run more instances
    of afl-fuzz than would be prudent (if you really want to).

  - `AFL_SKIP_CRASHES` causes AFL++ to tolerate crashing files in the input
    queue. This can help with rare situations where a program crashes only
    intermittently, but it's not really recommended under normal operating
    conditions.

  - Setting `AFL_HANG_TMOUT` allows you to specify a different timeout for
    deciding if a particular test case is a "hang". The default is 1 second
    or the value of the `-t` parameter, whichever is larger. Dialing the value
    down can be useful if you are very concerned about slow inputs, or if you
    don't want AFL++ to spend too much time classifying that stuff and just
    rapidly put all timeouts in that bin.

  - Setting `AFL_FORKSRV_INIT_TMOUT` allows you to specify a different timeout
    to wait for the forkserver to spin up. The default is the `-t` value times
    `FORK_WAIT_MULT` from `config.h` (usually 10), so for a `-t 100`, the
    default would wait for `1000` milliseconds. Setting a different time here is useful
    if the target has a very slow startup time, for example when doing
    full-system fuzzing or emulation, but you don't want the actual runs
    to wait too long for timeouts.

  - `AFL_NO_ARITH` causes AFL++ to skip most of the deterministic arithmetics.
    This can be useful to speed up the fuzzing of text-based file formats.

  - `AFL_NO_SNAPSHOT` will advice afl-fuzz not to use the snapshot feature
    if the snapshot lkm is loaded

  - `AFL_SHUFFLE_QUEUE` randomly reorders the input queue on startup. Requested
    by some users for unorthodox parallelized fuzzing setups, but not
    advisable otherwise.

  - `AFL_TMPDIR` is used to write the `.cur_input` file to if exists, and in
    the normal output directory otherwise. You would use this to point to
    a ramdisk/tmpfs. This increases the speed by a small value but also
    reduces the stress on SSDs.

  - When developing custom instrumentation on top of afl-fuzz, you can use
    `AFL_SKIP_BIN_CHECK` to inhibit the checks for non-instrumented binaries
    and shell scripts; and `AFL_DUMB_FORKSRV` in conjunction with the `-n`
    setting to instruct afl-fuzz to still follow the fork server protocol
    without expecting any instrumentation data in return.

  - When running in the `-M` or `-S` mode, setting `AFL_IMPORT_FIRST` causes the
    fuzzer to import test cases from other instances before doing anything

docs/Changelog.md
# Changelog

  This is the list of all noteworthy changes made in every public release of
  the tool. See README.md for the general instruction manual.

## Staying informed

Want to stay in the loop on major new features? Join our mailing list by
sending a mail to <afl-users+subscribe@googlegroups.com>.


### Version ++3.00a (develop)
  - llvm_mode/ and gcc_plugin/ moved to instrumentation/
  - all compilers combined to afl-cc which emulates the previous ones
  - afl-llvm/gcc-rt.o merged into afl-compiler-rt.o
  - afl-fuzz
    - memory limits are now disabled by default, set them with -m if required
    - deterministic fuzzing is now disabled by default and can be enabled with
      -D. It is still enabled by default for -M.
    - a new seed selection was implemented that uses weighted randoms based on
      a schedule performance score, which is much better that the previous
      walk the whole queue approach. Select the old mode with -Z (auto enabled
      with -M)
    - rpc.statsd support by Edznux, thanks a lot!
    - Marcel Boehme submitted a patch that improves all AFFast schedules :)
    - not specifying -M or -S will now auto-set "-S default"
    - reading testcases from -i now descends into subdirectories
    - allow up to 4 times the -x command line option
    - loaded extras now have a duplicate protection
    - If test cases are too large we do a partial read on the maximum
      supported size
    - longer seeds with the same trace information will now be ignored
      for fuzzing but still be used for splicing
    - crashing seeds are now not prohibiting a run anymore but are
      skipped. They are used for splicing though.
    - update MOpt for expanded havoc modes
    - added NO_SPLICING compile option and makefile define
    - added INTROSPECTION make target that writes all mutations to
      out/NAME/introspection.txt
    - print special compile time options used in help output
    - when using -c cmplog, one of the childs was not killed, fixed
    - somewhere we broke -n dumb fuzzing, fixed
    - new env. var. AFL_NO_COLOR (or AFL_NO_COLOUR) to suppress colored
      console output (when configured with USE_COLOR and not ALWAYS_COLORED)
  - instrumentation
    - We received an enhanced gcc_plugin module from AdaCore, thank you
      very much!!
    - not overriding -Ox or -fno-unroll-loops anymore
    - we now have our own trace-pc-guard implementation. It is the same as
      -fsanitize-coverage=trace-pc-guard from llvm 12, but: it is a) inline
      and b) works from llvm 10.0.1 + onwards :)
    - new llvm pass: dict2file via AFL_LLVM_DICT2FILE, create afl-fuzz
      -x dictionary of string comparisons found during compilation
    - LTO autodict now also collects interesting cmp comparisons,
      std::string compare + find + ==, bcmp
    - fix crash in dict2file for integers > 64 bit
  - custom mutators
    - added a new custom mutator: symcc -> https://github.com/eurecom-s3/symcc/
    - added a new custom mutator: libfuzzer that integrates libfuzzer mutations
    - Our afl++ Grammar-Mutator is now better integrated into custom_mutators/
    - added INTROSPECTION support for custom modules
    - python fuzz function was not optional, fixed
  - unicornafl synced with upstream (arm64 fix, better rust bindings)
  - renamed AFL_DEBUG_CHILD_OUTPUT to AFL_DEBUG_CHILD


### Version ++2.68c (release)
  - added the GSoC excellent afl++ grammar mutator by Shengtuo to our
    custom_mutators/ (see custom_mutators/README.md) - or get it here:
    https://github.com/AFLplusplus/Grammar-Mutator
  - a few QOL changes for Apple and its outdated gmake
  - afl-fuzz:
    - fix for auto dictionary entries found during fuzzing to not throw out
      a -x dictionary
    - added total execs done to plot file
    - AFL_MAX_DET_EXTRAS env variable added to control the amount of
      deterministic dict entries without recompiling.
    - AFL_FORKSRV_INIT_TMOUT env variable added to control the time to wait
      for the forkserver to come up without the need to increase the overall
      timeout.
    - bugfix for cmplog that results in a heap overflow based on target data
      (thanks to the magma team for reporting!)
    - write fuzzing setup into out/fuzzer_setup (environment variables and
      command line)
  - custom mutators:
    - added afl_custom_fuzz_count/fuzz_count function to allow specifying
      the number of fuzz attempts for custom_fuzz
  - llvm_mode:
    - ported SanCov to LTO, and made it the default for LTO. better
      instrumentation locations
    - Further llvm 12 support (fast moving target like afl++ :-) )
    - deprecated LLVM SKIPSINGLEBLOCK env environment


### Version ++2.67c (release)
  - Support for improved afl++ snapshot module:
    https://github.com/AFLplusplus/AFL-Snapshot-LKM
  - Due to the instrumentation needing more memory, the initial memory sizes
    for -m have been increased
  - afl-fuzz:
     - added -F option to allow -M main fuzzers to sync to foreign fuzzers,
       e.g. honggfuzz or libfuzzer
     - added -b option to bind to a specific CPU
     - eliminated CPU affinity race condition for -S/-M runs
     - expanded havoc mode added, on no cycle finds add extra splicing and
       MOpt into the mix
     - fixed a bug in redqueen for strings and made deterministic with -s
     - Compiletime autodictionary fixes
  - llvm_mode:
     - now supports llvm 12
     - support for AFL_LLVM_ALLOWLIST/AFL_LLVM_DENYLIST (previous
       AFL_LLVM_WHITELIST and AFL_LLVM_INSTRUMENT_FILE are deprecated and
       are matched to AFL_LLVM_ALLOWLIST). The format is compatible to llvm
       sancov, and also supports function matching :)
     - added neverzero counting to trace-pc/pcgard
     - fixes for laf-intel float splitting (thanks to mark-griffin for
       reporting)
     - fixes for llvm 4.0
     - skipping ctors and ifuncs for instrumentation
     - LTO: switch default to the dynamic memory map, set AFL_LLVM_MAP_ADDR

include/config.h
/*
   american fuzzy lop++ - vaguely configurable bits
   ------------------------------------------------

   Originally written by Michal Zalewski

   Now maintained by Marc Heuse <mh@mh-sec.de>,
                     Heiko Eißfeldt <heiko.eissfeldt@hexco.de>,
                     Andrea Fioraldi <andreafioraldi@gmail.com>,
                     Dominik Maier <mail@dmnk.co>

   Copyright 2016, 2017 Google Inc. All rights reserved.
   Copyright 2019-2020 AFLplusplus Project. All rights reserved.

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at:

     http://www.apache.org/licenses/LICENSE-2.0

 */

#ifndef _HAVE_CONFIG_H
#define _HAVE_CONFIG_H

#include "types.h"

/* Version string: */

// c = release, d = volatile github dev, e = experimental branch
#define VERSION "++3.00a"

/******************************************************
 *                                                    *
 *  Settings that may be of interest to power users:  *
 *                                                    *
 ******************************************************/

/* console output colors: There are three ways to configure its behavior
 * 1. default: colored outputs fixed on: defined USE_COLOR && defined ALWAYS_COLORED 
 *    The env var. AFL_NO_COLOR will have no effect
 * 2. defined USE_COLOR && !defined ALWAYS_COLORED
 *    -> depending on env var AFL_NO_COLOR=1 colors can be switched off
 *    at run-time. Default is to use colors.
 * 3. colored outputs fixed off: !defined USE_COLOR
 *    The env var. AFL_NO_COLOR will have no effect
*/

/* Comment out to disable terminal colors (note that this makes afl-analyze
   a lot less nice): */

#define USE_COLOR

#ifdef USE_COLOR
/* Comment in to always enable terminal colors */
/* Comment out to enable runtime controlled terminal colors via AFL_NO_COLOR */
#define ALWAYS_COLORED  1
#endif

/* StatsD config
   Config can be adjusted via AFL_STATSD_HOST and AFL_STATSD_PORT environment
   variable.
*/
#define STATSD_UPDATE_SEC 1
#define STATSD_DEFAULT_PORT 8125
#define STATSD_DEFAULT_HOST "127.0.0.1"

/* If you want to have the original afl internal memory corruption checks.
   Disabled by default for speed. it is better to use "make ASAN_BUILD=1". */

//#define _WANT_ORIGINAL_AFL_ALLOC

/* Comment out to disable fancy ANSI boxes and use poor man's 7-bit UI: */

#ifndef ANDROID_DISABLE_FANCY  // Fancy boxes are ugly from adb
  #define FANCY_BOXES
#endif

/* Default timeout for fuzzed code (milliseconds). This is the upper bound,
   also used for detecting hangs; the actual value is auto-scaled: */

#define EXEC_TIMEOUT 1000

/* Timeout rounding factor when auto-scaling (milliseconds): */

#define EXEC_TM_ROUND 20

/* 64bit arch MACRO */
#if (defined(__x86_64__) || defined(__arm64__) || defined(__aarch64__))
  #define WORD_SIZE_64 1
#endif

/* Default memory limit for child process (MB) 0 = disabled : */

#define MEM_LIMIT 0

/* Default memory limit when running in QEMU mode (MB) 0 = disabled : */

#define MEM_LIMIT_QEMU 0

/* Default memory limit when running in Unicorn mode (MB) 0 = disabled : */

#define MEM_LIMIT_UNICORN 0

/* Number of calibration cycles per every new test case (and for test
   cases that show variable behavior): */

#define CAL_CYCLES 8
#define CAL_CYCLES_LONG 40

/* Number of subsequent timeouts before abandoning an input file: */

#define TMOUT_LIMIT 250

/* Maximum number of unique hangs or crashes to record: */

#define KEEP_UNIQUE_HANG 500
#define KEEP_UNIQUE_CRASH 5000

/* Baseline number of random tweaks during a single 'havoc' stage: */

#define HAVOC_CYCLES 256
#define HAVOC_CYCLES_INIT 1024

/* Maximum multiplier for the above (should be a power of two, beware
   of 32-bit int overflows): */

#define HAVOC_MAX_MULT 64
#define HAVOC_MAX_MULT_MOPT 64

/* Absolute minimum number of havoc cycles (after all adjustments): */

#define HAVOC_MIN 12

/* Power Schedule Divisor */
#define POWER_BETA 1
#define MAX_FACTOR (POWER_BETA * 32)

/* Maximum stacking for havoc-stage tweaks. The actual value is calculated
   like this:

   n = random between 1 and HAVOC_STACK_POW2
   stacking = 2^n

   In other words, the default (n = 4) produces 2, 4, 8, 16
   stacked tweaks: */

#define HAVOC_STACK_POW2 4

/* Caps on block sizes for cloning and deletion operations. Each of these
   ranges has a 33% probability of getting picked, except for the first
   two cycles where smaller blocks are favored: */

#define HAVOC_BLK_SMALL 32
#define HAVOC_BLK_MEDIUM 128
#define HAVOC_BLK_LARGE 1500

/* Extra-large blocks, selected very rarely (<5% of the time): */

#define HAVOC_BLK_XL 32768

/* Probabilities of skipping non-favored entries in the queue, expressed as
   percentages: */

#define SKIP_TO_NEW_PROB 99     /* ...when there are new, pending favorites */
#define SKIP_NFAV_OLD_PROB 95   /* ...no new favs, cur entry already fuzzed */
#define SKIP_NFAV_NEW_PROB 75   /* ...no new favs, cur entry not fuzzed yet */

/* Splicing cycle count: */

#define SPLICE_CYCLES 15

/* Nominal per-splice havoc cycle length: */

#define SPLICE_HAVOC 32

/* Maximum offset for integer addition / subtraction stages: */

#define ARITH_MAX 35

/* Limits for the test case trimmer. The absolute minimum chunk size; and
   the starting and ending divisors for chopping up the input file: */

#define TRIM_MIN_BYTES 4
#define TRIM_START_STEPS 16
#define TRIM_END_STEPS 1024

/* Maximum size of input file, in bytes (keep under 100MB): */

#define MAX_FILE (1 * 1024 * 1024)

/* The same, for the test case minimizer: */

#define TMIN_MAX_FILE (10 * 1024 * 1024)

/* Block normalization steps for afl-tmin: */

#define TMIN_SET_MIN_SIZE 4
#define TMIN_SET_STEPS 128

/* Maximum dictionary token size (-x), in bytes: */

#define MAX_DICT_FILE 128

/* Length limits for auto-detected dictionary tokens: */

#define MIN_AUTO_EXTRA 3
#define MAX_AUTO_EXTRA 32

/* Maximum number of user-specified dictionary tokens to use in deterministic
   steps; past this point, the "extras/user" step will be still carried out,
   but with proportionally lower odds: */

#define MAX_DET_EXTRAS 256

/* Maximum number of auto-extracted dictionary tokens to actually use in fuzzing
   (first value), and to keep in memory as candidates. The latter should be much
   higher than the former. */

#define USE_AUTO_EXTRAS 128
#define MAX_AUTO_EXTRAS (USE_AUTO_EXTRAS * 64)

/* Scaling factor for the effector map used to skip some of the more
   expensive deterministic steps. The actual divisor is set to
   2^EFF_MAP_SCALE2 bytes: */

#define EFF_MAP_SCALE2 3

/* Minimum input file length at which the effector logic kicks in: */

#define EFF_MIN_LEN 128

/* Maximum effector density past which everything is just fuzzed
   unconditionally (%): */

#define EFF_MAX_PERC 90

/* UI refresh frequency (Hz): */

#define UI_TARGET_HZ 5

/* Fuzzer stats file and plot update intervals (sec): */

#define STATS_UPDATE_SEC 60
#define PLOT_UPDATE_SEC 5

/* Smoothing divisor for CPU load and exec speed stats (1 - no smoothing). */

#define AVG_SMOOTHING 16

/* Sync interval (every n havoc cycles): */

#define SYNC_INTERVAL 8

/* Output directory reuse grace period (minutes): */

#define OUTPUT_GRACE 25

/* Uncomment to use simple file names (id_NNNNNN): */

// #define SIMPLE_FILES

/* List of interesting values to use in fuzzing. */

#define INTERESTING_8                                    \
  -128,    /* Overflow signed 8-bit when decremented  */ \
      -1,  /*                                         */ \
      0,   /*                                         */ \
      1,   /*                                         */ \
      16,  /* One-off with common buffer size         */ \
      32,  /* One-off with common buffer size         */ \
      64,  /* One-off with common buffer size         */ \
      100, /* One-off with common buffer size         */ \
      127                        /* Overflow signed 8-bit when incremented  */

#define INTERESTING_8_LEN 9

#define INTERESTING_16                                    \
  -32768,   /* Overflow signed 16-bit when decremented */ \
      -129, /* Overflow signed 8-bit                   */ \
      128,  /* Overflow signed 8-bit                   */ \
      255,  /* Overflow unsig 8-bit when incremented   */ \
      256,  /* Overflow unsig 8-bit                    */ \
      512,  /* One-off with common buffer size         */ \
      1000, /* One-off with common buffer size         */ \
      1024, /* One-off with common buffer size         */ \
      4096, /* One-off with common buffer size         */ \
      32767                      /* Overflow signed 16-bit when incremented */

#define INTERESTING_16_LEN 10

#define INTERESTING_32                                          \
  -2147483648LL,  /* Overflow signed 32-bit when decremented */ \
      -100663046, /* Large negative number (endian-agnostic) */ \
      -32769,     /* Overflow signed 16-bit                  */ \
      32768,      /* Overflow signed 16-bit                  */ \
      65535,      /* Overflow unsig 16-bit when incremented  */ \
      65536,      /* Overflow unsig 16 bit                   */ \
      100663045,  /* Large positive number (endian-agnostic) */ \
      2147483647                 /* Overflow signed 32-bit when incremented */

#define INTERESTING_32_LEN 8

/***********************************************************
 *                                                         *
 *  Really exotic stuff you probably don't want to touch:  *
 *                                                         *
 ***********************************************************/

/* Call count interval between reseeding the libc PRNG from /dev/urandom: */

#define RESEED_RNG 100000

/* The default maximum testcase cache size in MB, 0 = disable.
   A value between 50 and 250 is a good default value. Note that the
   number of entries will be auto assigned if not specified via the
   AFL_TESTCACHE_ENTRIES env variable */

#define TESTCASE_CACHE_SIZE 50

/* Maximum line length passed from GCC to 'as' and used for parsing
   configuration files: */

#define MAX_LINE 8192

/* Environment variable used to pass SHM ID to the called program. */

#define SHM_ENV_VAR "__AFL_SHM_ID"

/* Environment variable used to pass SHM FUZZ ID to the called program. */

#define SHM_FUZZ_ENV_VAR "__AFL_SHM_FUZZ_ID"

/* Other less interesting, internal-only variables. */

#define CLANG_ENV_VAR "__AFL_CLANG_MODE"
#define AS_LOOP_ENV_VAR "__AFL_AS_LOOPCHECK"
#define PERSIST_ENV_VAR "__AFL_PERSISTENT"
#define DEFER_ENV_VAR "__AFL_DEFER_FORKSRV"

/* In-code signatures for deferred and persistent mode. */

#define PERSIST_SIG "##SIG_AFL_PERSISTENT##"
#define DEFER_SIG "##SIG_AFL_DEFER_FORKSRV##"

/* Distinctive bitmap signature used to indicate failed execution: */

#define EXEC_FAIL_SIG 0xfee1dead

/* Distinctive exit code used to indicate MSAN trip condition: */

#define MSAN_ERROR 86

/* Designated file descriptors for forkserver commands (the application will
   use FORKSRV_FD and FORKSRV_FD + 1): */

#define FORKSRV_FD 198

/* Fork server init timeout multiplier: we'll wait the user-selected
   timeout plus this much for the fork server to spin up. */

#define FORK_WAIT_MULT 10

/* Calibration timeout adjustments, to be a bit more generous when resuming
   fuzzing sessions or trying to calibrate already-added internal finds.
   The first value is a percentage, the other is in milliseconds: */

#define CAL_TMOUT_PERC 125
#define CAL_TMOUT_ADD 50

/* Number of chances to calibrate a case before giving up: */

#define CAL_CHANCES 3

/* Map size for the traced binary (2^MAP_SIZE_POW2). Must be greater than
   2; you probably want to keep it under 18 or so for performance reasons
   (adjusting AFL_INST_RATIO when compiling is probably a better way to solve
   problems with complex programs). You need to recompile the target binary
   after changing this - otherwise, SEGVs may ensue. */

#define MAP_SIZE_POW2 16
#define MAP_SIZE (1 << MAP_SIZE_POW2)

/* Maximum allocator request size (keep well under INT_MAX): */

#define MAX_ALLOC 0x40000000

/* A made-up hashing seed: */

#define HASH_CONST 0xa5b35705

/* Constants for afl-gotcpu to control busy loop timing: */

#define CTEST_TARGET_MS 5000
#define CTEST_CORE_TRG_MS 1000
#define CTEST_BUSY_CYCLES (10 * 1000 * 1000)

/* Enable NeverZero counters in QEMU mode */

#define AFL_QEMU_NOT_ZERO

/* AFL RedQueen */

#define CMPLOG_SHM_ENV_VAR "__AFL_CMPLOG_SHM_ID"

/* CPU Affinity lockfile env var */

#define CPU_AFFINITY_ENV_VAR "__AFL_LOCKFILE"

/* Uncomment this to use inferior block-coverage-based instrumentation. Note
   that you need to recompile the target binary for this to have any effect: */

// #define COVERAGE_ONLY

/* Uncomment this to ignore hit counts and output just one bit per tuple.
   As with the previous setting, you will need to recompile the target
   binary: */

// #define SKIP_COUNTS

/* Uncomment this to use instrumentation data to record newly discovered paths,
   but do not use them as seeds for fuzzing. This is useful for conveniently
   measuring coverage that could be attained by a "dumb" fuzzing algorithm: */

// #define IGNORE_FINDS

/* Text mutations */

/* Minimum length of a queue input to be evaluated for "is_ascii"? */

#define AFL_TXT_MIN_LEN 12

/* What is the minimum percentage of ascii characters present to be classifed
   as "is_ascii"? */

#define AFL_TXT_MIN_PERCENT 94

/* How often to perform ASCII mutations 0 = disable, 1-8 are good values */

#define AFL_TXT_BIAS 6

/* Maximum length of a string to tamper with */

#define AFL_TXT_STRING_MAX_LEN 1024

/* Maximum mutations on a string */

#define AFL_TXT_STRING_MAX_MUTATIONS 6

#endif                                                  /* ! _HAVE_CONFIG_H */

